{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "03733728",
   "metadata": {},
   "source": [
    "# Multi-Omic Data Analysis\n",
    "In the first part of this notebook, we will look at putting together all of the -omics that we have seen so far: genomics, transcriptomics, and proteomics. \n",
    "<br></br>\n",
    "In the second part of this notebook, your assignment is to put this into action and demonstrate a general knowledge of the different -omics using both TCGA and CPTAC data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d0279801",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "91ce2f2e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         \r"
     ]
    }
   ],
   "source": [
    "#cptac.download('Brca') # shouldn't need this if already downloaded\n",
    "brca = cptac.Brca()\n",
    "\n",
    "clinical = brca.get_clinical() # Clinical Data\n",
    "mutation = brca.get_somatic_mutation() # Genomic Data\n",
    "rna = brca.get_transcriptomics() # Transcriptomic Data\n",
    "protein = brca.get_proteomics() # Proteomic Data\n",
    "protein.columns = protein.columns.get_level_values(0) # Fix protein data columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e969ab21",
   "metadata": {},
   "source": [
    "## Let's revisit the structure of our different -omics...\n",
    "### Clinical"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36aed571",
   "metadata": {},
   "source": [
    "This is the data about the patient that was collected at the time that the sample was taken.\n",
    "Rows are patients, columns are variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3e20f16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Replicate_Measurement_IDs</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>TMT.Plex</th>\n",
       "      <th>TMT.Channel</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Ischemia.Time.in.Minutes</th>\n",
       "      <th>PAM50</th>\n",
       "      <th>NMF.Cluster</th>\n",
       "      <th>NMF.Cluster.Membership.Score</th>\n",
       "      <th>Age.in.Month</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>ER.Updated.Clinical.Status</th>\n",
       "      <th>PR.Clinical.Status</th>\n",
       "      <th>ERBB2.Updated.Clinical.Status</th>\n",
       "      <th>TNBC.Updated.Clinical.Status</th>\n",
       "      <th>ERBB2.Proteogenomic.Status</th>\n",
       "      <th>TOP2A.Proteogenomic.Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>CPT000814</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>13</td>\n",
       "      <td>127C</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>black.or.african.american</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>CPT001846</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>12</td>\n",
       "      <td>128C</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>white</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR001</th>\n",
       "      <td>X01BR001</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>2</td>\n",
       "      <td>129N</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.782</td>\n",
       "      <td>660.0</td>\n",
       "      <td>female</td>\n",
       "      <td>black.or.african.american</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR008</th>\n",
       "      <td>X01BR008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>127C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR009</th>\n",
       "      <td>X01BR009</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>16</td>\n",
       "      <td>127N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Basal</td>\n",
       "      <td>Basal-I</td>\n",
       "      <td>0.825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>positive</td>\n",
       "      <td>negative</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Replicate_Measurement_IDs Sample_Tumor_Normal TMT.Plex TMT.Channel  \\\n",
       "Patient_ID                                                                      \n",
       "CPT000814                  CPT000814               Tumor       13        127C   \n",
       "CPT001846                  CPT001846               Tumor       12        128C   \n",
       "X01BR001                    X01BR001               Tumor        2        129N   \n",
       "X01BR008                    X01BR008               Tumor       16        127C   \n",
       "X01BR009                    X01BR009               Tumor       16        127N   \n",
       "\n",
       "Name            Stage  Ischemia.Time.in.Minutes  PAM50 NMF.Cluster  \\\n",
       "Patient_ID                                                           \n",
       "CPT000814   Stage IIA                       NaN  Basal     Basal-I   \n",
       "CPT001846   Stage III                       NaN  Basal     Basal-I   \n",
       "X01BR001    Stage IIB                       0.0  Basal     Basal-I   \n",
       "X01BR008          NaN                       NaN  Basal     Basal-I   \n",
       "X01BR009          NaN                       NaN  Basal     Basal-I   \n",
       "\n",
       "Name        NMF.Cluster.Membership.Score  Age.in.Month  Gender  \\\n",
       "Patient_ID                                                       \n",
       "CPT000814                          1.000           NaN     NaN   \n",
       "CPT001846                          0.672           NaN     NaN   \n",
       "X01BR001                           0.782         660.0  female   \n",
       "X01BR008                           0.958           NaN     NaN   \n",
       "X01BR009                           0.825           NaN     NaN   \n",
       "\n",
       "Name                        Ethnicity ER.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                         \n",
       "CPT000814   black.or.african.american                   negative   \n",
       "CPT001846                       white                   negative   \n",
       "X01BR001    black.or.african.american                   negative   \n",
       "X01BR008                          NaN                        NaN   \n",
       "X01BR009                          NaN                   negative   \n",
       "\n",
       "Name       PR.Clinical.Status ERBB2.Updated.Clinical.Status  \\\n",
       "Patient_ID                                                    \n",
       "CPT000814            negative                           NaN   \n",
       "CPT001846            negative                           NaN   \n",
       "X01BR001             negative                      negative   \n",
       "X01BR008                  NaN                           NaN   \n",
       "X01BR009             negative                           NaN   \n",
       "\n",
       "Name       TNBC.Updated.Clinical.Status ERBB2.Proteogenomic.Status  \\\n",
       "Patient_ID                                                           \n",
       "CPT000814                      positive                   negative   \n",
       "CPT001846                      positive                   negative   \n",
       "X01BR001                       positive                   negative   \n",
       "X01BR008                            NaN                   negative   \n",
       "X01BR009                       positive                   negative   \n",
       "\n",
       "Name       TOP2A.Proteogenomic.Status  \n",
       "Patient_ID                             \n",
       "CPT000814                    negative  \n",
       "CPT001846                    negative  \n",
       "X01BR001                     negative  \n",
       "X01BR008                     negative  \n",
       "X01BR009                     negative  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d3e797f",
   "metadata": {},
   "source": [
    "### Mutation\n",
    "Our mutation data looks a little different than it did as a MAF file, but the important information is still here. Rows are patients, Gene is the name of the gene, Mutation is the type of mutation, and Location is where the mutation occurs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f87792d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>ABCG8</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.V660L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>ACTRT1</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.S138F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>ADA</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.A73T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>AHCTF1</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>p.I1012Sfs*9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>ANGPTL4</td>\n",
       "      <td>Silent</td>\n",
       "      <td>p.C341C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name           Gene           Mutation      Location\n",
       "Patient_ID                                          \n",
       "CPT000814     ABCG8  Missense_Mutation       p.V660L\n",
       "CPT000814    ACTRT1  Missense_Mutation       p.S138F\n",
       "CPT000814       ADA  Missense_Mutation        p.A73T\n",
       "CPT000814    AHCTF1    Frame_Shift_Del  p.I1012Sfs*9\n",
       "CPT000814   ANGPTL4             Silent       p.C341C"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aea8109",
   "metadata": {},
   "source": [
    "### RNA and Protein\n",
    "These both look exactly the same. The rows are patients, columns are genes, and the numbers represent log2 fold change compared to the standard. Note that the RNA DataFrame has roughly 2x as many columns as the Protein DataFrame. This is because of non-coding genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "033bcbd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>1.9265</td>\n",
       "      <td>2.4267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.3718</td>\n",
       "      <td>3.6437</td>\n",
       "      <td>-4.5301</td>\n",
       "      <td>-5.2608</td>\n",
       "      <td>0.6260</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.7000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.2268</td>\n",
       "      <td>5.1482</td>\n",
       "      <td>-0.0202</td>\n",
       "      <td>0.9270</td>\n",
       "      <td>5.1699</td>\n",
       "      <td>3.1333</td>\n",
       "      <td>0.9835</td>\n",
       "      <td>5.9499</td>\n",
       "      <td>2.8499</td>\n",
       "      <td>3.6497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>3.6578</td>\n",
       "      <td>2.6524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.3253</td>\n",
       "      <td>-5.1069</td>\n",
       "      <td>-0.7692</td>\n",
       "      <td>5.6915</td>\n",
       "      <td>3.2060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.3115</td>\n",
       "      <td>...</td>\n",
       "      <td>3.2532</td>\n",
       "      <td>3.9191</td>\n",
       "      <td>-1.4924</td>\n",
       "      <td>-0.3262</td>\n",
       "      <td>3.2163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0909</td>\n",
       "      <td>7.0671</td>\n",
       "      <td>2.0780</td>\n",
       "      <td>3.2274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR001</th>\n",
       "      <td>0.9896</td>\n",
       "      <td>2.6363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.4734</td>\n",
       "      <td>-1.5052</td>\n",
       "      <td>-2.0918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.8816</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.3434</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9491</td>\n",
       "      <td>2.2365</td>\n",
       "      <td>0.1694</td>\n",
       "      <td>1.3074</td>\n",
       "      <td>2.8806</td>\n",
       "      <td>0.5653</td>\n",
       "      <td>1.5682</td>\n",
       "      <td>4.3852</td>\n",
       "      <td>2.1573</td>\n",
       "      <td>2.1127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR008</th>\n",
       "      <td>0.5535</td>\n",
       "      <td>2.2119</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.8701</td>\n",
       "      <td>3.2708</td>\n",
       "      <td>-1.9902</td>\n",
       "      <td>-5.0414</td>\n",
       "      <td>0.6957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.3095</td>\n",
       "      <td>...</td>\n",
       "      <td>3.1575</td>\n",
       "      <td>5.7822</td>\n",
       "      <td>-1.3608</td>\n",
       "      <td>0.3015</td>\n",
       "      <td>3.2620</td>\n",
       "      <td>1.7578</td>\n",
       "      <td>0.2989</td>\n",
       "      <td>5.4697</td>\n",
       "      <td>3.1783</td>\n",
       "      <td>3.1939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR009</th>\n",
       "      <td>2.8359</td>\n",
       "      <td>3.3449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.9307</td>\n",
       "      <td>2.6488</td>\n",
       "      <td>-3.1664</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9784</td>\n",
       "      <td>-5.7975</td>\n",
       "      <td>4.6341</td>\n",
       "      <td>...</td>\n",
       "      <td>2.7678</td>\n",
       "      <td>3.5378</td>\n",
       "      <td>-0.5147</td>\n",
       "      <td>0.2798</td>\n",
       "      <td>3.0986</td>\n",
       "      <td>-3.2737</td>\n",
       "      <td>1.3036</td>\n",
       "      <td>5.0219</td>\n",
       "      <td>3.6868</td>\n",
       "      <td>2.8524</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name          A1BG  A1BG-AS1  A1CF     A2M   A2ML1   A2MP1  A3GALT2  A4GALT  \\\n",
       "Patient_ID                                                                    \n",
       "CPT000814   1.9265    2.4267   NaN  5.3718  3.6437 -4.5301  -5.2608  0.6260   \n",
       "CPT001846   3.6578    2.6524   NaN  5.3253 -5.1069 -0.7692   5.6915  3.2060   \n",
       "X01BR001    0.9896    2.6363   NaN  5.4734 -1.5052 -2.0918      NaN  1.8816   \n",
       "X01BR008    0.5535    2.2119   NaN  5.8701  3.2708 -1.9902  -5.0414  0.6957   \n",
       "X01BR009    2.8359    3.3449   NaN  6.9307  2.6488 -3.1664      NaN  1.9784   \n",
       "\n",
       "Name         A4GNT    AAAS  ...  ZWILCH   ZWINT    ZXDA    ZXDB    ZXDC  \\\n",
       "Patient_ID                  ...                                           \n",
       "CPT000814      NaN  4.7000  ...  3.2268  5.1482 -0.0202  0.9270  5.1699   \n",
       "CPT001846      NaN  4.3115  ...  3.2532  3.9191 -1.4924 -0.3262  3.2163   \n",
       "X01BR001       NaN  3.3434  ...  1.9491  2.2365  0.1694  1.3074  2.8806   \n",
       "X01BR008       NaN  4.3095  ...  3.1575  5.7822 -1.3608  0.3015  3.2620   \n",
       "X01BR009   -5.7975  4.6341  ...  2.7678  3.5378 -0.5147  0.2798  3.0986   \n",
       "\n",
       "Name        ZYG11A  ZYG11B     ZYX   ZZEF1    ZZZ3  \n",
       "Patient_ID                                          \n",
       "CPT000814   3.1333  0.9835  5.9499  2.8499  3.6497  \n",
       "CPT001846      NaN  2.0909  7.0671  2.0780  3.2274  \n",
       "X01BR001    0.5653  1.5682  4.3852  2.1573  2.1127  \n",
       "X01BR008    1.7578  0.2989  5.4697  3.1783  3.1939  \n",
       "X01BR009   -3.2737  1.3036  5.0219  3.6868  2.8524  \n",
       "\n",
       "[5 rows x 23121 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "055471ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAED1</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMDC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN31</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>-0.6712</td>\n",
       "      <td>-0.2075</td>\n",
       "      <td>2.7959</td>\n",
       "      <td>1.3969</td>\n",
       "      <td>-1.0899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.6708</td>\n",
       "      <td>-0.3484</td>\n",
       "      <td>-0.4756</td>\n",
       "      <td>-0.7299</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.2868</td>\n",
       "      <td>-0.6536</td>\n",
       "      <td>0.3384</td>\n",
       "      <td>2.1169</td>\n",
       "      <td>1.3910</td>\n",
       "      <td>-2.1230</td>\n",
       "      <td>0.9136</td>\n",
       "      <td>-0.8082</td>\n",
       "      <td>-1.4793</td>\n",
       "      <td>0.9136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>1.3964</td>\n",
       "      <td>1.3302</td>\n",
       "      <td>-5.0948</td>\n",
       "      <td>0.7674</td>\n",
       "      <td>-1.6845</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.1022</td>\n",
       "      <td>-0.5814</td>\n",
       "      <td>0.2916</td>\n",
       "      <td>-2.2857</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.7592</td>\n",
       "      <td>0.4711</td>\n",
       "      <td>0.6018</td>\n",
       "      <td>0.2062</td>\n",
       "      <td>-0.2137</td>\n",
       "      <td>-2.1219</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>2.5814</td>\n",
       "      <td>-0.2852</td>\n",
       "      <td>-0.1074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR001</th>\n",
       "      <td>2.0219</td>\n",
       "      <td>1.6269</td>\n",
       "      <td>-3.2943</td>\n",
       "      <td>0.3352</td>\n",
       "      <td>-1.0739</td>\n",
       "      <td>1.2255</td>\n",
       "      <td>0.2754</td>\n",
       "      <td>-1.1187</td>\n",
       "      <td>-0.0534</td>\n",
       "      <td>-0.2519</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2306</td>\n",
       "      <td>-0.3010</td>\n",
       "      <td>0.3395</td>\n",
       "      <td>-0.5316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4996</td>\n",
       "      <td>0.7622</td>\n",
       "      <td>-1.5607</td>\n",
       "      <td>0.0256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR008</th>\n",
       "      <td>-0.5290</td>\n",
       "      <td>0.3267</td>\n",
       "      <td>1.4342</td>\n",
       "      <td>0.4938</td>\n",
       "      <td>-2.8676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0691</td>\n",
       "      <td>-0.3643</td>\n",
       "      <td>-1.8173</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.1789</td>\n",
       "      <td>0.2695</td>\n",
       "      <td>0.1506</td>\n",
       "      <td>1.0498</td>\n",
       "      <td>0.7546</td>\n",
       "      <td>1.7889</td>\n",
       "      <td>-0.2499</td>\n",
       "      <td>-0.2590</td>\n",
       "      <td>-0.1263</td>\n",
       "      <td>0.3725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X01BR009</th>\n",
       "      <td>1.2556</td>\n",
       "      <td>3.4489</td>\n",
       "      <td>2.8043</td>\n",
       "      <td>-0.2956</td>\n",
       "      <td>-1.7261</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.0471</td>\n",
       "      <td>-0.3547</td>\n",
       "      <td>-0.8298</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.3990</td>\n",
       "      <td>-0.2596</td>\n",
       "      <td>0.1898</td>\n",
       "      <td>-0.5010</td>\n",
       "      <td>-0.4189</td>\n",
       "      <td>0.3080</td>\n",
       "      <td>0.5057</td>\n",
       "      <td>0.2181</td>\n",
       "      <td>-0.2288</td>\n",
       "      <td>-0.2750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name          A1BG     A2M   A2ML1    AAAS    AACS   AADAT   AAED1   AAGAB  \\\n",
       "Patient_ID                                                                   \n",
       "CPT000814  -0.6712 -0.2075  2.7959  1.3969 -1.0899     NaN  1.6708 -0.3484   \n",
       "CPT001846   1.3964  1.3302 -5.0948  0.7674 -1.6845     NaN  2.1022 -0.5814   \n",
       "X01BR001    2.0219  1.6269 -3.2943  0.3352 -1.0739  1.2255  0.2754 -1.1187   \n",
       "X01BR008   -0.5290  0.3267  1.4342  0.4938 -2.8676     NaN     NaN -1.0691   \n",
       "X01BR009    1.2556  3.4489  2.8043 -0.2956 -1.7261     NaN     NaN -2.0471   \n",
       "\n",
       "Name          AAK1   AAMDC  ...  ZSCAN31  ZSWIM8    ZW10  ZWILCH   ZWINT  \\\n",
       "Patient_ID                  ...                                            \n",
       "CPT000814  -0.4756 -0.7299  ...  -5.2868 -0.6536  0.3384  2.1169  1.3910   \n",
       "CPT001846   0.2916 -2.2857  ...  -0.7592  0.4711  0.6018  0.2062 -0.2137   \n",
       "X01BR001   -0.0534 -0.2519  ...      NaN  0.2306 -0.3010  0.3395 -0.5316   \n",
       "X01BR008   -0.3643 -1.8173  ...  -2.1789  0.2695  0.1506  1.0498  0.7546   \n",
       "X01BR009   -0.3547 -0.8298  ...  -2.3990 -0.2596  0.1898 -0.5010 -0.4189   \n",
       "\n",
       "Name          ZXDC  ZYG11B     ZYX   ZZEF1    ZZZ3  \n",
       "Patient_ID                                          \n",
       "CPT000814  -2.1230  0.9136 -0.8082 -1.4793  0.9136  \n",
       "CPT001846  -2.1219  0.0860  2.5814 -0.2852 -0.1074  \n",
       "X01BR001       NaN  0.4996  0.7622 -1.5607  0.0256  \n",
       "X01BR008    1.7889 -0.2499 -0.2590 -0.1263  0.3725  \n",
       "X01BR009    0.3080  0.5057  0.2181 -0.2288 -0.2750  \n",
       "\n",
       "[5 rows x 10107 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "920d2cb7",
   "metadata": {},
   "source": [
    "## Reformatting data\n",
    "### Finding shared patients\n",
    "All 4 DataFrames we have here share the common characteristic that the rows are the patient_IDs. This is actually really helpful! However, not all DataFrames have the same number of patients. For example, the mutation DataFrame has many times more rows than patients because each patient shows up everytime for a new mutation. Othertimes, you have masked out some patients from a DataFrame but not the other. In order to find the shared patients between DataFrames, we can use the np.intersect1d() function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61ec006f",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# skeleton code\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m patients_df1 \u001b[38;5;241m=\u001b[39m \u001b[43mdf1\u001b[49m\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;66;03m# the indices of each DF are the patients within that DF\u001b[39;00m\n\u001b[1;32m      4\u001b[0m patients_df2 \u001b[38;5;241m=\u001b[39m df2\u001b[38;5;241m.\u001b[39mindex\n\u001b[1;32m      6\u001b[0m shared_patients \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mintersect1d(patients_df1, patients_df2) \u001b[38;5;66;03m# find shared and save indices\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df1' is not defined"
     ]
    }
   ],
   "source": [
    "# skeleton code\n",
    "\n",
    "patients_df1 = df1.index # the indices of each DF are the patients within that DF\n",
    "patients_df2 = df2.index\n",
    "\n",
    "shared_patients = np.intersect1d(patients_df1, patients_df2) # find shared and save indices\n",
    "\n",
    "shared_df1 = df1.loc[shared_patients, :] # use our list of shared patients to grab only those patients\n",
    "shared_df2 = df2.loc[shared_patients, :] # notice we are grabbing all columns and assigning to new df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32c8561c",
   "metadata": {},
   "source": [
    "Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aed1e92",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_MUC16_prot = protein[protein.loc[:, 'MUC16'] > 1]\n",
    "high_MUC16_RNA = rna[rna.loc[:, 'MUC16'] > 1]\n",
    "\n",
    "shared_MUC16 = np.intersect1d(high_MUC16_prot.index, high_MUC16_RNA.index)\n",
    "\n",
    "print(f'prot rows: {len(high_MUC16_prot)}')\n",
    "print(f'RNA rows: {len(high_MUC16_RNA)}')\n",
    "print(f'shared patients: {len(shared_MUC16)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "481ee29b",
   "metadata": {},
   "source": [
    "### Counting # of Somatic Mutations in Individual\n",
    "Say we wanted to look at how the number of mutations in MUC16 relate to RNA and protein expression. How do we do this? We can't use the prior approach since the Mutation DF has many more rows than # of patients. We want to get a count of the number of MUC16 mutations in each patient. We can use the  Series.value_counts() method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ab0e722",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Start by getting only mutations you are interested in\n",
    "MUC16_mutat_mask = mutation.loc[:, 'Gene'] == 'MUC16'\n",
    "MUC16_mutations = mutation.loc[MUC16_mutat_mask, :]\n",
    "MUC16_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d3d67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Then get a list of the patient names (hint: the index)\n",
    "patients_with_mutation = MUC16_mutations.index\n",
    "\n",
    "# Then use the value_counts() function to count how many times each patient appears\n",
    "MUC16_counts = patients_with_mutation.value_counts()\n",
    "print(MUC16_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afcb338c",
   "metadata": {},
   "source": [
    "Wow it looks like one patient, X01BR043, has the majority of all MUC16 mutations. Let's see what fraction of our MUC16 patients are in our shared high MUC16 RNA and Protein list from before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5f813b",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.intersect1d(MUC16_counts.index, shared_MUC16)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3618d0c2",
   "metadata": {},
   "source": [
    "Only two patients were shared between the list. Perhaps this means that the converse is true.. that mutations are linked to lower gene expression. Let's put all of this data together for our analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88904775",
   "metadata": {},
   "outputs": [],
   "source": [
    "patients_df = pd.DataFrame(\n",
    "    columns=['MUC16_Mutations', 'RNA_Expression', 'Protein_Expression'], \n",
    "    index=clinical.index)\n",
    "\n",
    "patients_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3323288",
   "metadata": {},
   "source": [
    "Now that we have our DataFrame, let's fill in the values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06d9aebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "patients_df.loc[MUC16_counts.index, 'MUC16_Mutations'] = MUC16_counts\n",
    "patients_df.loc[patients_df.index.difference(MUC16_counts.index), 'MUC16_Mutations'] = 0\n",
    "# df.index.difference(list) gives you all of the indices for the rows that are not in that list\n",
    "# here, we are setting all of the patients who don't appear in MUC16_Mutations to have 0 mutations\n",
    "\n",
    "patients_df.loc[:, 'RNA_Expression'] = rna.loc[:, 'MUC16']\n",
    "patients_df.loc[:, 'Protein_Expression'] = protein.loc[:, 'MUC16']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ec4ead1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(patients_df.isna().sum()) # count number of NAs in each column, let's drop the rows with NAs\n",
    "patients_df.dropna(axis=0, inplace=True) # drops rows with NA values, alternatively use axis=1 to drop columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac0145ad",
   "metadata": {},
   "source": [
    "Let's see if we can gain anything from a UMAP of our data..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14255265",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "scaler = StandardScaler()\n",
    "from umap import UMAP\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as lines\n",
    "\n",
    "reducer = UMAP()\n",
    "transformed_df = scaler.fit_transform(patients_df)\n",
    "embedding = reducer.fit_transform(transformed_df)\n",
    "\n",
    "colors_rna = np.where(patients_df.loc[:, 'RNA_Expression'] > 1, 'blue', 'grey')\n",
    "colors_protein = np.where(patients_df.loc[:, 'Protein_Expression'] > 1, 'red', 'grey')\n",
    "show_mutations = np.where(patients_df.loc[:, 'MUC16_Mutations'] > 0, 0.3, 0)\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(embedding[:, 0], embedding[:, 1], c=colors_rna, s=20, alpha=0.6)\n",
    "ax.scatter(embedding[:, 0], embedding[:, 1], c=colors_protein, s=20, alpha=0.5)\n",
    "ax.scatter(embedding[:, 0], embedding[:, 1], alpha=show_mutations, color='white', s=100, edgecolor='Black')\n",
    "\n",
    "rna = mpatches.Patch(color='blue', label='High MUC16 RNA')\n",
    "protein = mpatches.Patch(color='red', label='High MUC16 Protein')\n",
    "shared = mpatches.Patch(color='mediumorchid', label='High MUC16 RNA+Protein')\n",
    "mutations = lines.Line2D([], [], color=\"white\", alpha=0.5, marker='o', markersize=10, markerfacecolor=\"white\", markeredgecolor='black', label='MUC16 Mutation +')\n",
    "\n",
    "ax.legend(handles=[rna, protein, shared, mutations], bbox_to_anchor=[0.85,-0.05], fancybox=True, ncol=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afb792c2",
   "metadata": {},
   "source": [
    "From our plot, we can see that while the Mutation+ patients tend to be spread throughout the different categories, it doesn't seem like many are high in both MUC16 RNA expression AND MUC16 Protein expression."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93eff5d6",
   "metadata": {},
   "source": [
    "## Assignment\n",
    "This assignment has 2 parts and unlike the past few assignments, it will be more open-ended. Grading will be based on effort and thoughtful analysis, not whether your data was significant."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecad6d0e",
   "metadata": {},
   "source": [
    "**Exercise 1** - Investigate the Central Dogma (5pt)\n",
    "<br>\n",
    "Complete a multi-omic analysis analysis of a gene, looking at how mutation rate, RNA expression, and protein expression are correlated (or not!). Choose your gene logically and include a justification (e.g. relative protein abundance, , high nonsense mutation rate, based in the literature as having implications with breast cancer, etc)\n",
    "You may not use any of the following genes: TP53, TTN, PIK3CA, MUC16, BRCA1/2\n",
    "<br> </br>\n",
    "Include the following (1pt each):\n",
    "+ Justification for choosing the gene (e.g. relevant code used to pick or link to research article). One to two sentence explanation of the function of the gene.\n",
    "+ Examination of the somatic mutations with respect to type, number per individual, or density\n",
    "+ Some form of figure that shows relationship between RNA expression and Protein Expression\n",
    "+ A statistical test that gives a p-value. Interpret the p-value in terms of the test and the data. We've mostly looked at correlation tests but you don't necessarily have to use one. Scipy.stats is a great resource https://docs.scipy.org/doc/scipy/reference/stats.html#correlation-functions\n",
    "+ Three to four sentence writeup of your thoughts behind why you got the results you got (just brainstorm some ideas, it doesn't need to be super thorough)\n",
    "\n",
    "Remember: If there is no correlation, state that, you don't need to try multiple genes if you aren't getting promising results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882a9ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4e2205b",
   "metadata": {},
   "source": [
    "**Exercise 2** - Looking for Connections (5pt)\n",
    "As the types or size of data increases, it can be harder to find patterns within the data. This presents a perfect opportunity to use one of our newest techniques: machine learning. In this exercise, you will attempt to predict a clinical variable based on genetic and mutation data. \n",
    "<br></br>\n",
    "Our clinical data from CPTAC is pretty sparse however, and doesn't give us much in the way to predict. Luckily, we can use our TCGA data just as easily. Using the pd.read_csv() functions, read in your clinical, mutation, and transcriptomic data. \n",
    "<br></br>\n",
    "Your task is as follows (1pt each):\n",
    "+ Load in your TCGA analysis_data and choose one of the following clinical variables:\n",
    " vital_status, days_to_last_known_alive/days_to_death/days_to_last_followup (combine), histological_type, age_at_initial_pathologic_diagnosis, lymph_node_examined_count or breast_carcinoma_progesterone_receptor_status.\n",
    "+ Determine which genes you want to use for your regressor/classifier (you can use ones we've looked at, genes which are commonly associated with breast cancer outcomes, the most mutated genes, etc) and create a DataFrame that includes those genes RNA and mutation data. You can choose which data/how many data dimensions you want.\n",
    "+ Train a regressor/classifier (your choice of which method) on your chosen data with the target being your chosen variable (hint: if you choose a categorical variable, you will need to use a LabelEncoder())\n",
    "+ Measure and interpret the accuracy of the model. If you do not get above a 0.5 accuracy (for classification) or within 10% of the target (regression), retry with a different ML model, train test split, or genes used. If you cannot meet any of the goals within 5 tries, move on (it's okay!).\n",
    "+ Write a few sentences about why a machine learning algorithm like a regressor/classifier would be important in a clinical setting. Depending on whether you got it within 5 tries or not, discuss why you think your genes/parameters were/were not able to predict it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91195909",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write code here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
